Emit events outside of db/transaction

dpsutton opened the pull request:
## Emit events outside of db/transaction

When emitting events (particularly revision events), the event handler
loads the item from the db to ensure that all post-processing is
done. If done from inside a transaction, there's a race between the
event handler looking up the object from the db and the transaction
completing to put the item in the db. When the event handler wins, the
following code fails:

```clojure
(push-revision! :entity       Dashboard,
                :id           id,
                :object       (Dashboard id), ;; this is nil
                :user-id      user-id,
                :is-creation? (= :dashboard-create topic)
                :message      revision-message)
```

and then the schema check for `push-revision!` fails `(map? object)`
and there is no event recorded for creation. On edit, it might get the
pre-edit version and record an incorrect message or not create the
revision for the latest change, preventing going back in the future.

I'm only doing this in dashboard but these changes should be audited
and applied everywhere

dpsutton committed:
Emit events outside of db/transaction

When emitting events (particularly revision events), the event handler
loads the item from the db to ensure that all post-processing is
done. If done from inside a transaction, there's a race between the
event handler looking up the object from the db and the transaction
completing to put the item in the db. When the event handler wins, the
following code fails:

```clojure
(push-revision! :entity       Dashboard,
                :id           id,
                :object       (Dashboard id), ;; this is nil
                :user-id      user-id,
                :is-creation? (= :dashboard-create topic)
                :message      revision-message)
```

and then the schema check for `push-revision!` fails `(map? object)`
and there is no event recorded for creation. On edit, it might get the
pre-edit version and record an incorrect message or not create the
revision for the latest change, preventing going back in the future.

I'm only doing this in dashboard but these changes should be audited
and applied everywhere

nemanjaglumac commented:
Unfortunately, I can't review clojure code but tests seem to be passing so it the race condition should be fixed. I tried it locally and did a stress-test of the previously flaking Cypress test. All green :)
![image](https://user-images.githubusercontent.com/31325167/117024190-0e7ff380-acfa-11eb-9ffd-f4aafc2fee0c.png)

camsaul commented on src/metabase/models/dashboard_card.clj:
```diff
@@ -109,23 +109,23 @@
+                                                  :dashboardcard_id id
+                                                  {:order-by [[:position :asc]]})))
+       (update-dashboard-card-series! dashboard-card series)))
+    (u/prog1 (retrieve-dashboard-card id)
+      ;; todo: should this have an actor_id from the api user?
+      (events/publish-event! :dashboard-card-update <>))))
```
Do you still need the `prog1` here?

dpsutton replied:
yes if `<>` needs an actor_id assoc'd to it and not otherwise. Good catch. I'm _pretty_ sure `:actor_id` needs to be on there but it wasn't in the past.

dpsutton replied:
So interestingly, there is no event handler for `:dashboard-card-update` or `:dashboard-card-create`. And there hasn't been for the five years of its existence. I'm not sure how dashboard-cards and dashboards interact for the purposes of revisions. I wonder if we can properly revert them in all cases. Either way those events can be removed for now.

flamber replied:
We cannot revert properly for all revisions - sometimes fails, other time creates some weird corruption - that's issue 6884

camsaul reviewed (approved):
Awesome, LGTM

dpsutton committed:
Remove `:dashboard-card-update` and `:dashboard-card-create` events

These events were introduced 5 years ago and never have had an event
handler. No worries about race conditions when emitting them from a
transaction if nothing ever cared in the first place.
